felix/ipsets/nft_sets.go

```diff
@@ -0,0 +1,35 @@
+package ipsets
+
+import (
+	"fmt"
+
+	"github.com/projectcalico/calico/felix/ip"
+)
+
+// Constants for nftables, which uses alternative set types as well as
+// additional set types that are not supported by iptables.
+const (
+	NFTSetTypeAddr     = "ipv%d_addr"
```

Comment:
I ended up needing to make new set types for nftables because all of the parsing / canonization of members is keyed off of the type, so a "hash:net" needs different logic than the equivalent nftables set does (as their members are formatted differently).

I agree in general that I'm not really happy with how I've got the sets stuff laid out right now - it feels pretty icky. Perhaps I need to either. The alternative I saw was to make the calc graph aware of iptabes vs. nftables, and that also felt wrong. The way I have be thinking of it is that the "generic" calc graph interface just happens to be the same types / strings as the iptables implementation :sweat_smile:

Reply:
The factoring that occurs to me is to change this method
```
func (t IPSetType) CanonicaliseMember(member string) IPSetMember
```
into function (or a method of a dataplane-specific object):
```
func CanonicaliseMember(t IPSetType, member string) IPSetMember
```
and then have two copies, one for iptables and one for nftables?